pubsub/natspubsub/nats.go

```diff
@@ -0,0 +1,292 @@
+// ugorji driver to encode and decode driver.Message to []byte.
+//
+// As
+//
+// natspubsub exposes the following types for As:
+//  - Topic: *nats.Conn
+//  - Subscription: *nats.Subscription
+//  - Message: *nats.Msg
+
+package natspubsub // import "gocloud.dev/pubsub/natspubsub"
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/nats-io/go-nats"
+	"github.com/ugorji/go/codec"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+var errNotInitialized = errors.New("natspubsub: topic not initialized")
+
+type topic struct {
+	nc   *nats.Conn
+	subj string
+}
+
+// For encoding we use msgpack from github.com/ugorji/go.
+// It was already imported in go-cloud and is reasonably performant.
+// However the more recent version has better resource handling with
+// the addition of the following:
+// 	mh.ExplicitRelease = true
+// 	defer enc.Release()
+// However this is not compatible with etcd at the moment.
+// https://github.com/etcd-io/etcd/pull/10337
+var mh codec.MsgpackHandle
+
+func init() {
+	// driver.Message.Metadata type
+	dm := driver.Message{}
+	mh.MapType = reflect.TypeOf(dm.Metadata)
+}
+
+// We define our own version of message here for encoding that
+// only encodes Body and Metadata. Otherwise we would have to
+// add codec decorations to driver.Message.
+type encMsg struct {
+	Body     []byte            `codec:",omitempty"`
+	Metadata map[string]string `codec:",omitempty"`
+}
+
+// CreateTopic returns a *pubsub.Topic for use with NATS.
+// We delay checking for the proper syntax here.
+func CreateTopic(nc *nats.Conn, topicName string) *pubsub.Topic {
+	return pubsub.NewTopic(createTopic(nc, topicName))
+}
+
+// createTopic returns the driver for CreateTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func createTopic(nc *nats.Conn, topicName string) driver.Topic {
+	return &topic{nc, topicName}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
+	if t == nil || t.nc == nil {
+		return errNotInitialized
+	}
+
+	// Reuse if possible.
+	var em encMsg
+	var raw [1024]byte
+	b := raw[:0]
+	enc := codec.NewEncoderBytes(&b, &mh)
+
+	for _, m := range msgs {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		enc.ResetBytes(&b)
+		em.Body, em.Metadata = m.Body, m.Metadata
+		if err := enc.Encode(em); err != nil {
+			return err
+		}
+		if err := t.nc.Publish(t.subj, b); err != nil {
+			return err
+		}
+	}
+	// Per specification this is supposed to only return after
+	// a message has been sent. Normally NATS is very efficient
+	// at sending messages in batches on its own and also handles
+	// disconnected buffering during a reconnect event. We will
+	// let NATS handle this for now. If needed we could add a
+	// FlushWithContext() call which ensures the connected server
+	// has processed all the messages.
+	return nil
```

Comment:
Does `FlushWithContext` return an error? I'm a bit concerned that we're returning success here when messages may not have actually been sent, which is misleading. I'd rather start off correct and slow....

Reply:
Flush in the NATS system sends a PING to the server and waits for a PONG, meaning all protocol messages sent will have been processed by the server. Makes things very slow and we usually discourage its use since NATS core is an at-most-once system. But even in STAN (at-least-once) we do not use it and wait on a reply from the stan-server.

Reply:
Let me know if we need to iterate some more, but again for a system that is at-most-once this makes sense.

Reply:
Agreed.